src/Nethermind/Nethermind.Evm/Tracing/GasEstimator.cs

```diff
@@ -51,24 +50,26 @@ public long Estimate(Transaction tx, BlockHeader header, EstimateGasTracer gasTr
                 ? tx.GasLimit
                 : header.GasLimit;
 
-            UInt256 senderBalance = _stateProvider.GetBalance(tx.SenderAddress);
-
-            // Calculate and return additional gas required in case of insufficient funds.
-            if (tx.Value != UInt256.Zero && tx.Value >= senderBalance)
+            // Execute binary search to find the optimal gas estimation.
+            try
             {
-                return gasTracer.CalculateAdditionalGasRequired(tx, releaseSpec);
+                return BinarySearchEstimate(leftBound, rightBound, tx, header, cancellationToken)
+                       + Math.Max(0, gasTracer.CalculateAdditionalGasRequired(tx, releaseSpec));
```

Comment:
There is a chance (looks more like a bug) for the `AdditionalGasRequired` to be `< 0`, so we ignore it in those cases.

Reply:
Shouldn't this be handled inside the tracer?

Reply:
Why this is even added? I don't see this in the original code, if we estimate with actual running this shouldn't be needed and probably makes the tests behave differently in the first place?

Reply:
> Why this is even added?

I had to add this after running the whole test suite and seeing that there were failures that weren't flaky tests: there are integrations tests that verify the gas estimation (which have been updated on this PR) that fail if we don't include this `AdditionalGasRequired`.